ruby_programming/git/advanced_git_lesson.md

```diff
@@ -0,0 +1,182 @@
+
+~~~bash
+pick c85c71dd5 Rewrite Introduction
+pick 4ad2c8df6 Add information on using amend
+pick 729db1da4 Begin work on Remotes
+~~~
+
+If I wanted to edit one of them, I would change the word `pick` to be `edit` for the appropriate commit. If I wanted to remove a commit I would simply remove it from that list, and if I wanted to change the order I would move them around on the list. Let's see what all three would look like! 
+
+~~~bash
+pick 4ad2c8df6 Add information on using amend
+edit c85c71dd5 Rewrite Introduction
+~~~
+
+This would allow me to edit the `Rewrite Introduction` commit, switch its order with the `Add information on using amend` commit, and completely remove the most recent `Begin work on Remotes` commit. When I'm certain I want to do this, I'll save and exit the editor, which will allow me to edit my commit with the following instructions:
+
+~~~bash
+You can amend the commit now, with
+       git commit --amend
+Once youre satisfied with your changes, run
+       git rebase --continue
+~~~
+
+These commit changes would be disasterous for me, so I won't *actually* do it, but it's a great tool to keep in mind for your future projects. That's all there is to it! It seems simple, but this is a very dangerous tool if misused, so be careful. Most importantly, remember to **only rebase commits that exist in your repository, never rebase commits others may work off of.**
+
+
+## Squashing Commits
+
+Using `squash` for your commits is a very handy way of keeping your git history tidy. Squashing makes it easier for others to understand the history of your project. This will become especially important when you begin to merge features in larger projects. What often happens when a feature is merged, is you end up with some visually complex logs of all the changes a feature branch had on a main branch. These commits are important while the feature is in development, but aren't really necessary when looking through the entire history of your main branch.
+
+Let's say I want to `squash` all three of the commits I previously worked with together into the first commit on the list, which is `Rewrite Introduction`. I would simply  `pick` that commit as the commit that the other two are being `squash`ed into:
+
+~~~bash
+pick c85c71dd5 Rewrite Introduction
+squash 4ad2c8df6 Add information on using amend
+squash 729db1da4 Begin work on Remotes
+~~~
+
+That's it! When I save and exit the editor, the bottom two commits on that list get squashed into the top commit. 
+
+## Splitting Up a Commit
+
+Before diving into Remotes, we're going to have a look at a handy git command called `reset`. Let's say one of my commits was `c85c71dd5 Add knight class and Add attack module`. That's a mouthful, and a bit much for one commit as you should have learned in the previous lesson on commits. So what we're going to do is split it up into two smaller commits, also using the interactive `rebase` tool. 
```

Comment:
I think that if we're touching on `reset` it might be good to mention `reset --hard`, why it's dangerous, and the difference between a hard reset and a soft reset.